In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/test-file/tested.csv
/kaggle/input/titanic/train.csv
/kaggle/input/titanic/test.csv
/kaggle/input/titanic/gender_submission.csv


# Problem Statement :

* Use the Titanic dataset to build a model that predicts whether a
passenger on the Titanic survived or not. This is a classic beginner
project with readily available data.

* The dataset typically used for this project contains information
about individual passengers, such as their age, gender, ticket
class, fare, cabin, and whether or not they survived.

 # Loading the Data
 
 * **train data** -> train.csv
 * **test data** -> tested.csv

In [3]:
trainDataTit = pd.read_csv("/kaggle/input/titanic/train.csv")
print(trainDataTit.head)

<bound method NDFrame.head of      PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                     

In [4]:
testDataTit = pd.read_csv("/kaggle/input/test-file/tested.csv")
print(testDataTit.head())

   PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S  


# Checking % of *Men* and *Women* Survived

To check how many **females** and **males** *survived* during the titanic sink.

In [5]:
womenSurvived = trainDataTit.loc[trainDataTit.Sex == 'female']["Survived"]
rateWomenSurvived = sum(womenSurvived)/len(womenSurvived)
print("Percentage of women who survived : ", rateWomenSurvived)

Percentage of women who survived :  0.7420382165605095


In [6]:
menSurvived = trainDataTit.loc[trainDataTit.Sex == 'male']["Survived"]
rateMenSurvived = sum(menSurvived)/len(menSurvived)
print("Percentage of men who survived : ", rateMenSurvived)

Percentage of men who survived :  0.18890814558058924


**Conclusion :**
As we can see, *survival rate of WOMEN is **more** than MEN*, but it is not okay to say that all WOMEN survived and all MEN die.

# Selecting the Model :

Now, Applying ML Model - Here we use *Supervised Model* - (**Random Forest Model)**
* Splitting the data into training and testing sets
* Evaluatinig the Model to check accuracy score



In [7]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

y = trainDataTit["Survived"]

features = ["Pclass", "Sex", "SibSp", "Parch"] # these features are taken to be in consideration for the prediction
x = pd.get_dummies(trainDataTit[features])
# x_test = pd.get_dummies(testDataTit[features])

model = RandomForestClassifier(n_estimators=100, max_depth=5, random_state=3)
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=42)

model.fit(x_train, y_train)
y_pred = model.predict(x_test)



/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [8]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred ))

              precision    recall  f1-score   support

           0       0.81      0.87      0.83       105
           1       0.79      0.70      0.74        74

    accuracy                           0.80       179
   macro avg       0.80      0.78      0.79       179
weighted avg       0.80      0.80      0.80       179



In [9]:
testDataTit.head()
X_test = pd.get_dummies(testDataTit[features])

In [10]:
prediction=model.predict(X_test)

In [11]:
output = pd.DataFrame({'PassengerId': testDataTit.PassengerId, 'Survived': prediction})
output.to_csv('submission.csv', index=False)
print("Your submission was successfully saved!")

Your submission was successfully saved!


# Discussion

The Titanic survival prediction project used **RandomForestClassifier Model** technique to anticipate passenger survival. 8Data preprocessing, feature engineering, and model selection* were pivotal. Feature importance analysis unveiled factors influencing survival. The project highlights the potential of machine learning in historical event analysis. By predicting survival outcomes, this effort contributes to understanding and commemorating the Titanic tragedy while showcasing the power of predictive analytics in complex scenarios.

# Conclusion

In this project, we employed Supervised Model analysis and Machine Learning techniques to effectively anticipate passenger survival. By navigating data preprocessing, feature engineering, and model selection, the project achieved **accurate prediction of 0.80**.
The effort sheds light on the tragic events of the Titanic while showcasing the value of machine learning in historical analysis. This project not only provides insights into the factors that influenced survival but also underscores the potential of predictive modeling in understanding complex real-world scenarios.